cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterServiceLookup.java

```diff
@@ -178,20 +180,23 @@ private EndpointStrategy discoverService(CacheKey key)
   private EndpointStrategy discoverDefaultService(CacheKey key)
     throws UnsupportedEncodingException, ExecutionException {
     // Try only path routing
-    String lookupService = genLookupName(key.getService(), DEFAULT_SERVICE_NAME, key.getFirstPathPart());
-    return discover(lookupService);
+    String lookupService = genLookupName(key.getRouteDestination().getServiceName(),
+                                         DEFAULT_SERVICE_NAME, key.getFirstPathPart());
+    return discover(new RouteDestination(lookupService));
   }
 
-  private EndpointStrategy discover(String discoverName) throws ExecutionException {
-    LOG.debug("Looking up service name {}", discoverName);
+  private EndpointStrategy discover(RouteDestination routeDestination) throws ExecutionException {
+    LOG.debug("Looking up service name {}", routeDestination);
     // If its a user service, then use DistributionEndpoint Strategy
-    ServiceDiscovered serviceDiscovered = discoveryServiceClient.discover(discoverName);
+    String serviceName = routeDestination.getServiceName();
+    ServiceDiscovered serviceDiscovered = discoveryServiceClient.discover(serviceName);
 
-    EndpointStrategy endpointStrategy = ServiceDiscoverable.isServiceDiscoverable(discoverName) ?
-      new DistributionEndpointStrategy(serviceDiscovered, routeStore, ServiceDiscoverable.getId(discoverName)) :
+    EndpointStrategy endpointStrategy = ServiceDiscoverable.isServiceDiscoverable(serviceName) ?
+      new DistributionEndpointStrategy(serviceDiscovered, routeStore, ServiceDiscoverable.getId(serviceName),
```

Comment:
I see. Let's call it `UserServiceEndpointStrategy`, and also call `payloadFilter` as `version` then.